File: php/src/SinglyLinkList.php
<?php


namespace DataStructure;


class SinglyLinkList
{

    private ?SinglyLinkListNode $head = null;
    private ?SinglyLinkListNode $tail = null;
    private int $size = 0;

    public function size(): int
    {
        return $this->size;
    }

    public function tail(): ?SinglyLinkListNode
    {
        return $this->tail;
    }

    public function add($value)
    {
        $node = new SinglyLinkListNode($value);
        $this->size++;
        if (! $this->head)
        {
            $this->head = $node;
            $this->tail = $this->head;
            return;
        }

        $this->tail->next = $node;
        $this->tail = $node;
    }

    public function remove($value)
    {
        $node = $this->head();
        $left = null;

        while ($node)
        {
            if ($value == $node->value)
            {
                if (! $left)
                {
                    $this->head = $node->next;
                }
                else
                {
                    $left->next = $node->next;
                }
                $this->size--;
                break;
            }

            $left = $node;
            $node = $node->next;
        }
    }

    public function head(): ?SinglyLinkListNode
    {
        return $this->head;
    }

    public function find($value): ?SinglyLinkListNode
    {
        $node = $this->head();

        while ($node)
        {
            if ($value == $node->value)
            {
                return $node;
            }

            $node = $node->next;
        }

        return null;
    }
}

File: python3/singly_link_list.py
from singly_link_list_node import SinglyLinkListNode


class SinglyLinkList:
    head: SinglyLinkListNode = None
    tail: SinglyLinkListNode = None
    __size: int = 0

    @property
    def size(self):
        return self.__size

    def add(self, value):
        node = SinglyLinkListNode(value)
        self.__size += 1
        if self.head is None:
            self.head = node
            self.tail = node
            return

        self.tail.next = node
        self.tail = node

    def remove(self, value):
        node: SinglyLinkListNode = self.head
        left: SinglyLinkListNode = None
        self.__size -= 1
        while node is not None:

            if value == node.value:
                if left is None:
                    self.head = node.next
                else:
                    left.next = node.next

            left = node
            node = node.next
